File: python/polymorphisme/src/enseignant.py
from src.personne import Personne


class Enseignant(Personne):
    def __init__(self, nom: str, age: int, matiere: str, salaire: float):
        super().__init__(nom, age)
        self.matiere = matiere
        self.salaire = salaire

    @property
    def matiere(self) -> str:
        return self.__matiere

    @matiere.setter
    def matiere(self, valeur: str):
        if valeur is None or not valeur.strip():
            raise ValueError("La matière ne peut pas être vide.")
        self.__matiere = valeur.strip()

    @property
    def salaire(self) -> float:
        return self.__salaire

    @salaire.setter
    def salaire(self, valeur: float):
        if valeur < 0:
            raise ValueError("Le salaire ne peut pas être négatif.")
        self.__salaire = valeur

    def afficher_details(self) -> str:
        return (
            f"Enseignant: {self.nom}, {self.age} ans, "
            f"Matière: {self.matiere}, "
            f"Salaire: {self.salaire}"
        )

    def __str__(self):
        return self.afficher_details()

    def __repr__(self):
        return (
            f"Enseignant({self.nom!r}, {self.age!r}, "
            f"{self.matiere!r}, {self.salaire!r})"
        )

File: java/src/main/java/polymorphisme/Enseignant.java
package polymorphisme;

public class Enseignant extends Personne {
    private String matiere;
    private double salaire;

    public Enseignant(String nom, int age, String matiere, double salaire) {
        super(nom, age);
        setMatiere(matiere);
        setSalaire(salaire);
    }

    public String getMatiere() {
        return matiere;
    }

    public void setMatiere(String matiere) {
        if (matiere == null || matiere.trim().isEmpty()) {
            throw new IllegalArgumentException("La matière ne peut pas être vide.");
        }
        this.matiere = matiere.trim();
    }

    public double getSalaire() {
        return salaire;
    }

    public void setSalaire(double salaire) {
        if (salaire < 0) {
            throw new IllegalArgumentException("Le salaire ne peut pas être négatif.");
        }
        this.salaire = salaire;
    }

    @Override
    public String afficherDetails() {
        return "Enseignant: " + getNom()
                + ", " + getAge() + " ans"
                + ", Matière: " + matiere
                + ", Salaire: " + salaire;
    }

    @Override
    public String toString() {
        return afficherDetails();
    }
}